sdk/identity/src/token_credentials/client_certificate_credentials.rs

```diff
@@ -0,0 +1,231 @@
+    OpensslError(ErrorStack),
+    #[error("HTTP operation failed {0}")]
+    ReqwestError(reqwest::Error),
+}
+
+#[derive(Deserialize, Debug, Default)]
+#[serde(default)]
+struct AadTokenResponse {
+    token_type: String,
+    expires_in: u64,
+    ext_expires_in: u64,
+    access_token: String,
+}
+
+#[async_trait::async_trait]
+impl TokenCredential for ClientCertificateCredential {
+    type Error = ClientCertificateCredentialError;
+
+    async fn get_token(&self, resource: &str) -> Result<TokenResponse, Self::Error> {
+        let options = self.options();
+        let url = &format!(
+            "{}/{}/oauth2/v2.0/token",
+            options.authority_host(),
+            self.tenant_id
+        );
+
+        let certificate = base64::decode(&self.client_certificate)
+            .map_err(ClientCertificateCredentialError::DecodeError)?;
+        let certificate = Pkcs12::from_der(&certificate)
+            .map_err(ClientCertificateCredentialError::OpensslError)?
+            .parse(&self.client_certificate_pass)
+            .map_err(ClientCertificateCredentialError::OpensslError)?;
+
+        let thumbprint = ClientCertificateCredential::get_thumbprint(&certificate.cert)
+            .map_err(ClientCertificateCredentialError::OpensslError)?;
+
+        let uuid = uuid::Uuid::new_v4();
+        let current_time = Utc::now().timestamp();
+        let expiry_time = current_time + DEFAULT_REFRESH_TIME;
+        let x5t = base64::encode(&thumbprint);
+
+        let header = match options.send_certificate_chain {
+            true => format!(
+                r#"{{"alg":"RS256","typ":"JWT", "x5t":"{}", "x5c":["{}"]}}"#,
+                x5t, self.client_certificate
+            ),
+            false => format!(r#"{{"alg":"RS256","typ":"JWT", "x5t":"{}"}}"#, x5t),
+        };
+        let header = ClientCertificateCredential::as_jwt_part(header.as_bytes());
+
+        let payload = format!(
+            r#"{{"aud":"{}","exp":{},"iss": "{}", "jti": "{}", "nbf": {}, "sub": "{}"}}"#,
+            url, expiry_time, self.client_id, uuid, current_time, self.client_id
+        );
+        let payload = ClientCertificateCredential::as_jwt_part(payload.as_bytes());
+
+        let jwt = format!("{}.{}", header, payload);
+        let signature = ClientCertificateCredential::sign(&jwt, &certificate.pkey)
+            .map_err(ClientCertificateCredentialError::OpensslError)?;
+        let sig = ClientCertificateCredential::as_jwt_part(&signature);
+        let client_assertion = format!("{}.{}", jwt, sig);
+
+        let form_data = vec![
+            ("client_id", self.client_id.to_owned()),
+            ("scope", format!("{}/.default", resource)),
+            (
+                "client_assertion_type",
+                "urn:ietf:params:oauth:client-assertion-type:jwt-bearer".to_owned(),
+            ),
+            ("client_assertion", client_assertion),
+            ("grant_type", "client_credentials".to_owned()),
+        ];
```

Comment:
@schaabs any concerns with this particular bit?

Reply:
@chlowell since Scott is OOF, what do you think about the substance of this method? You (and Scott) know that area of identity far better than I.

Reply:
Sounds like this is fine for now.